std/math/fp2.asm

```diff
@@ -0,0 +1,159 @@
+use std::math::ff::inverse;
+use std::convert::fe;
+use std::convert::int;
+use std::convert::expr;
+use std::field::modulus;
+use std::prover::eval;
+
+// An extension field element, represented as a tuple (a0, a1) of expressions (which evaluate to field elements).
+// An element of F_{p^2} is a polynomial a0 + a1 * x (mod x^2 - 7).
+// The polynomial x^2 - 7 is irreducible in both the Goldilocks and BN254 field
+// (which follows from 7 not having a square root in both fields).
+enum Fp2Expr {
+    Fp2(expr, expr)
+}
+
+// An element of F_{p^2}, but represented as a tuple of *evaluated* field elements.
+enum Fp2Value {
```

Comment:
Why not use a generic enum called
```
enum QuadraticPoly<T> {
  P(T, T)
}
```
?

Reply:
Ah no, it's not just a quadratic poly, it's a quadratic poly modulo `x^2 - 7`.

Reply:
Nice, that worked! I now just called it `Fp2<T>`.